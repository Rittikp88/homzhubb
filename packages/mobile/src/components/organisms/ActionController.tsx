import React from 'react';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { RecordAssetActions } from '@homzhub/common/src/modules/recordAsset/actions';
import { RecordAssetSelectors } from '@homzhub/common/src/modules/recordAsset/selectors';
import LeaseTermController from '@homzhub/mobile/src/components/organisms/LeaseTermController';
import { SaleTermController } from '@homzhub/mobile/src/components/organisms/SaleTermController';
import { ManageTermController } from '@homzhub/mobile/src/components/organisms/ManageTermController';
import { Country } from '@homzhub/common/src/domain/models/CountryCode';
import { TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { AssetGroupTypes } from '@homzhub/common/src/constants/AssetGroup';
import { FurnishingTypes } from '@homzhub/common/src/constants/Terms';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import { IState } from '@homzhub/common/src/modules/interfaces';

interface IProps {
  typeOfPlan: TypeOfPlan;
  isSplitAsUnits: boolean;
  onNextStep: () => void;
  country: Country;
  assetGroupType: string;
  furnishing: FurnishingTypes;
  lastVisitedStep: ILastVisitedStep;
}

interface IDispatchProps {
  setTermId: (termId: number) => void;
  getMaintenanceUnits: () => void;
}

interface IStateProps {
  currentTermId: number;
  currentAssetId: number;
}

type Props = IStateProps & IDispatchProps & IProps;

// TODO (Aditya): Optimize, look at un-required props drilling
class ActionController extends React.PureComponent<Props, {}> {
  public componentDidMount = (): void => {
    const { getMaintenanceUnits, assetGroupType } = this.props;
    if (assetGroupType === AssetGroupTypes.COM) {
      getMaintenanceUnits();
    }
  };

  public render = (): React.ReactNode => {
    const { onNextStep } = this.props;
    const {
      typeOfPlan,
      currentTermId,
      setTermId,
      currentAssetId,
      country: { currencies },
      assetGroupType,
      lastVisitedStep,
      furnishing,
    } = this.props;
    return (
      <>
        {typeOfPlan === TypeOfPlan.SELL && (
          <SaleTermController
            currentAssetId={currentAssetId}
            assetGroupType={assetGroupType}
            currencyData={currencies[0]}
            onNextStep={this.onNextStep}
          />
        )}
        {typeOfPlan === TypeOfPlan.RENT && (
          <LeaseTermController
            currentAssetId={currentAssetId}
            currentTermId={currentTermId}
            assetGroupType={assetGroupType}
            setTermId={setTermId}
            furnishing={furnishing}
            lastVisitedStep={lastVisitedStep}
            currencyData={currencies[0]}
            onNextStep={onNextStep}
          />
        )}
        {typeOfPlan === TypeOfPlan.MANAGE && (
          <ManageTermController
            currentAssetId={currentAssetId}
            assetGroupType={assetGroupType}
            currencyData={currencies[0]}
            onNextStep={this.onNextStep}
          />
        )}
      </>
    );
  };

  private onNextStep = async (type: TypeOfPlan): Promise<void> => {
    const { onNextStep, lastVisitedStep, currentAssetId } = this.props;

    const last_visited_step = {
      ...lastVisitedStep,
      listing: {
        ...lastVisitedStep.listing,
        type,
        is_listing_created: true,
      },
    };
    await AssetRepository.updateAsset(currentAssetId, { last_visited_step });
    onNextStep();
  };
}

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { setTermId, getMaintenanceUnits } = RecordAssetActions;
  return bindActionCreators({ setTermId, getMaintenanceUnits }, dispatch);
};

const mapStateToProps = (state: IState): IStateProps => {
  const { getCurrentTermId, getCurrentAssetId } = RecordAssetSelectors;
  return {
    currentTermId: getCurrentTermId(state),
    currentAssetId: getCurrentAssetId(state),
  };
};

const HOC = connect(mapStateToProps, mapDispatchToProps)(ActionController);
export { HOC as ActionController };
