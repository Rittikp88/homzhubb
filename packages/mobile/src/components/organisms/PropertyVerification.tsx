import React from 'react';
import { StyleSheet, TouchableOpacity, View } from 'react-native';
import { WithTranslation, withTranslation } from 'react-i18next';
import DocumentPicker from 'react-native-document-picker';
import ImagePicker from 'react-native-image-crop-picker';
import { cloneDeep, findIndex } from 'lodash';
import { ObjectMapper } from '@homzhub/common/src/utils/ObjectMapper';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { PlatformUtils } from '@homzhub/common/src/utils/PlatformUtils';
import {
  AllowedAttachmentFormats,
  AttachmentError,
  AttachmentService,
  AttachmentType,
} from '@homzhub/common/src/services/AttachmentService';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { theme } from '@homzhub/common/src/styles/theme';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import Selfie from '@homzhub/common/src/assets/images/selfie.svg';
import { Button } from '@homzhub/common/src/components/atoms/Button';
import { ImageThumbnail } from '@homzhub/common/src/components/atoms/ImageThumbnail';
import { Label, Text } from '@homzhub/common/src/components/atoms/Text';
import { UploadBox } from '@homzhub/common/src/components/molecules/UploadBox';
import {
  ExistingVerificationDocuments,
  IExistingVerificationDocuments,
  IPostVerificationDocuments,
  VerificationDocumentCategory,
  VerificationDocumentTypes,
} from '@homzhub/common/src/domain/models/VerificationDocuments';
import { selfieInstruction } from '@homzhub/common/src/constants/AsssetVerification';
import { TypeOfPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import { IUpdateAssetParams } from '@homzhub/common/src/domain/repositories/interfaces';

interface IPropertyVerificationState {
  verificationTypes: VerificationDocumentTypes[];
  existingDocuments: ExistingVerificationDocuments[];
  localDocuments: ExistingVerificationDocuments[];
  isLoading: boolean;
}

interface IProps {
  typeOfPlan: TypeOfPlan;
  updateStep: () => void;
  propertyId: number;
  lastVisitedStep: ILastVisitedStep;
}

type Props = WithTranslation & IProps;

export class PropertyVerification extends React.PureComponent<Props, IPropertyVerificationState> {
  public state = {
    verificationTypes: [],
    existingDocuments: [],
    localDocuments: [],
    isLoading: false,
  };

  public componentDidMount = async (): Promise<void> => {
    const { propertyId } = this.props;
    await this.getVerificationTypes();
    await this.getExistingDocuments(propertyId);
  };

  public render(): React.ReactElement {
    const { t } = this.props;
    const { verificationTypes, existingDocuments, localDocuments, isLoading } = this.state;
    const totalDocuments = existingDocuments.concat(localDocuments);
    return (
      <>
        <View style={styles.container}>{this.renderVerificationTypes()}</View>
        <Button
          type="primary"
          title={t('common:continue')}
          disabled={totalDocuments.length < verificationTypes.length || isLoading}
          containerStyle={styles.buttonStyle}
          onPress={this.postPropertyVerificationDocuments}
        />
      </>
    );
  }

  public renderVerificationTypes = (): React.ReactNode => {
    const { verificationTypes } = this.state;
    return verificationTypes.map((verificationType: VerificationDocumentTypes, index: number) => {
      const data: VerificationDocumentTypes = verificationType;
      return (
        <View style={styles.proofChild} key={index}>
          <Text type="small" textType="semiBold" style={styles.title}>
            {data.title}
          </Text>
          {verificationType.name === VerificationDocumentCategory.SELFIE_ID_PROOF ? (
            <>
              <Selfie style={styles.selfie} />
              {selfieInstruction.map((instruction, i) => {
                return (
                  <Label type="regular" textType="regular" style={styles.instruction} key={i}>
                    {instruction}
                  </Label>
                );
              })}
            </>
          ) : (
            <>
              {data.description !== '' && (
                <Label type="regular" textType="regular" style={styles.subTitle}>
                  {data.description}
                </Label>
              )}
            </>
          )}
          {this.renderImageOrUploadBox(verificationType)}
        </View>
      );
    });
  };

  public renderImageOrUploadBox = (currentData: VerificationDocumentTypes): React.ReactElement => {
    const { existingDocuments, localDocuments } = this.state;

    const onPress = async (): Promise<void> => this.handleVerificationDocumentUploads(currentData);

    const totalDocuments = existingDocuments.concat(localDocuments);
    const thumbnailIndex = findIndex(totalDocuments, (document: ExistingVerificationDocuments) => {
      return currentData.id === document.verificationDocumentType.id;
    });

    if (thumbnailIndex !== -1) {
      const currentDocument: ExistingVerificationDocuments = totalDocuments[thumbnailIndex];
      const thumbnailImage = currentDocument.document.link;
      const fileType = currentDocument.document.link.split('/')?.pop()?.split('.');

      const onDeleteImageThumbnail = (): Promise<void> =>
        this.deleteDocument(currentDocument, currentDocument.isLocalDocument);

      const { mimeType } = currentDocument.document;

      return mimeType === AllowedAttachmentFormats.AppPdf || !fileType || fileType[1] === 'pdf' ? (
        <View style={styles.pdfContainer}>
          <Text type="small" textType="regular" style={styles.pdfName}>
            {currentDocument.document.name || fileType || [0]}
          </Text>
          <TouchableOpacity style={styles.iconContainer} onPress={onDeleteImageThumbnail}>
            <Icon name={icons.close} size={22} color={theme.colors.shadow} />
          </TouchableOpacity>
        </View>
      ) : (
        <ImageThumbnail imageUrl={thumbnailImage} onIconPress={onDeleteImageThumbnail} />
      );
    }

    return (
      <UploadBox
        icon={currentData.icon}
        header={currentData.label}
        subHeader={currentData.helpText}
        onPress={onPress}
        containerStyle={styles.uploadBox}
      />
    );
  };

  public uploadDocument = async (verificationDocumentId: number, data: VerificationDocumentTypes): Promise<void> => {
    const document = await DocumentPicker.pick({
      type: [DocumentPicker.types.allFiles],
    });
    if (Object.values(AllowedAttachmentFormats).includes(document.type)) {
      const source = { uri: document.uri, type: document.type, name: document.name };
      this.updateLocalDocuments(verificationDocumentId, source, data);
    } else {
      AlertHelper.error({ message: data.helpText });
    }
  };

  public captureSelfie = (verificationDocumentId: number, data: VerificationDocumentTypes): void => {
    ImagePicker.openCamera({
      compressImageMaxWidth: 400,
      compressImageMaxHeight: 400,
      compressImageQuality: PlatformUtils.isAndroid() ? 1 : 0.8,
      useFrontCamera: true,
    }).then((image: any) => {
      const source = {
        uri: image.path,
        type: image.mime,
        name: PlatformUtils.isIOS() ? image.filename : image.path.substring(image.path.lastIndexOf('/') + 1),
      };
      this.updateLocalDocuments(verificationDocumentId, source, data);
    });
  };

  public handleVerificationDocumentUploads = async (data: VerificationDocumentTypes): Promise<void> => {
    const verificationDocumentId = data.id;
    const verificationDocumentType = data.name;
    if (verificationDocumentType === VerificationDocumentCategory.SELFIE_ID_PROOF) {
      this.captureSelfie(verificationDocumentId, data);
    } else {
      await this.uploadDocument(verificationDocumentId, data);
    }
  };

  public getVerificationTypes = async (): Promise<void> => {
    const { typeOfPlan } = this.props;
    try {
      const response: VerificationDocumentTypes[] = await AssetRepository.getVerificationDocumentTypes();
      const filteredResponse = response.filter((data: VerificationDocumentTypes) => {
        return data.category === typeOfPlan || data.category === 'IDENTITY';
      });
      this.setState({
        verificationTypes: filteredResponse,
      });
    } catch (error) {
      AlertHelper.error({ message: error.message });
    }
  };

  public getExistingDocuments = async (propertyId: number): Promise<void> => {
    try {
      let existingDocuments: ExistingVerificationDocuments[] = [];
      existingDocuments = await AssetRepository.getExistingVerificationDocuments(propertyId);
      if (existingDocuments.length === 0) {
        existingDocuments = await AssetRepository.getAssetIdentityDocuments();
      }
      this.setState({
        existingDocuments,
      });
    } catch (error) {
      AlertHelper.error({ message: error.message });
    }
  };

  public postPropertyVerificationDocuments = async (): Promise<void> => {
    const { propertyId, updateStep, t, lastVisitedStep, typeOfPlan } = this.props;
    const { localDocuments, existingDocuments } = this.state;

    const updateAssetPayload: IUpdateAssetParams = {
      last_visited_step: {
        ...lastVisitedStep,
        listing: {
          ...lastVisitedStep.listing,
          type: typeOfPlan,
          is_verification_done: true,
        },
      },
    };

    if (localDocuments.length === 0) {
      await AssetRepository.updateAsset(propertyId, updateAssetPayload);
      updateStep();
      return;
    }
    const formData = new FormData();
    localDocuments.forEach((document: ExistingVerificationDocuments) => {
      formData.append('files[]', {
        // @ts-ignore
        name: document.document.name,
        uri: document.document.link,
        // @ts-ignore
        type: document.document.mimeType,
      });
    });

    this.setState({ isLoading: true });

    try {
      const response = await AttachmentService.uploadImage(formData, AttachmentType.ASSET_VERIFICATION);

      const { data } = response;
      const postRequestBody: IPostVerificationDocuments[] = [];

      localDocuments.forEach((document: ExistingVerificationDocuments, index: number) => {
        postRequestBody.push({
          verification_document_type_id: document.verificationDocumentType.id,
          document_id: data[index].id,
        });
      });

      existingDocuments.forEach((document: ExistingVerificationDocuments) => {
        if (!document.id) {
          postRequestBody.push({
            verification_document_type_id: document.verificationDocumentType.id,
            document_id: document.document.id,
          });
        }
      });
      this.setState({ isLoading: false });

      await AssetRepository.postVerificationDocuments(propertyId, postRequestBody);
      await this.getExistingDocuments(propertyId);
      await AssetRepository.updateAsset(propertyId, updateAssetPayload);
      updateStep();
    } catch (e) {
      this.setState({ isLoading: false });
      if (e === AttachmentError.UPLOAD_IMAGE_ERROR) {
        AlertHelper.error({ message: t('common:fileCorrupt') });
      }
    }
  };

  public deleteDocument = async (
    document: ExistingVerificationDocuments,
    isLocalDocument: boolean | null
  ): Promise<void> => {
    const { existingDocuments, localDocuments, verificationTypes, isLoading } = this.state;
    const clonedDocuments = isLocalDocument ? cloneDeep(localDocuments) : cloneDeep(existingDocuments);
    if (!document.id) {
      const documentIndex = findIndex(clonedDocuments, (existingDocument: ExistingVerificationDocuments) => {
        return existingDocument.verificationDocumentType.id === document.verificationDocumentType.id;
      });
      if (documentIndex !== -1) {
        clonedDocuments.splice(documentIndex, 1);
        const key = isLocalDocument ? 'localDocuments' : 'existingDocuments';
        this.setState({
          existingDocuments,
          localDocuments,
          verificationTypes,
          [key]: clonedDocuments,
          isLoading,
        });
      }
    } else {
      const { propertyId } = this.props;
      try {
        await AssetRepository.deleteVerificationDocument(propertyId, document.id);
        await this.getExistingDocuments(propertyId);
      } catch (error) {
        AlertHelper.error({ message: error.message });
      }
    }
  };

  public updateLocalDocuments = (
    verificationDocumentId: number,
    source: { uri: string; type: string; name: string },
    data: VerificationDocumentTypes
  ): void => {
    const imageObject: IExistingVerificationDocuments = {
      id: null,
      verification_document_type: ObjectMapper.serialize(data),
      document: {
        id: data.id,
        name: source.name,
        attachment_type: source.type,
        mime_type: source.type,
        link: source.uri,
      },
      is_local_document: true,
    };
    const { localDocuments } = this.state;
    this.setState({
      localDocuments: [...localDocuments, ObjectMapper.deserialize(ExistingVerificationDocuments, imageObject)],
    });
  };
}

export default withTranslation(LocaleConstants.namespacesKey.property)(PropertyVerification);

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 16,
    marginTop: 4,
    backgroundColor: theme.colors.white,
  },
  proofChild: {
    marginBottom: 10,
    marginTop: 20,
  },
  uploadBox: {
    marginTop: 20,
  },
  title: {
    color: theme.colors.darkTint4,
  },
  subTitle: {
    color: theme.colors.darkTint3,
    marginVertical: 10,
  },
  pdfContainer: {
    flex: 0,
    flexDirection: 'row',
    backgroundColor: theme.colors.white,
    padding: 16,
    borderColor: theme.colors.primaryColor,
    borderWidth: 1,
    borderStyle: 'solid',
    marginTop: 10,
    borderRadius: 4,
  },
  iconContainer: {
    width: 30,
    height: 30,
    borderRadius: 30,
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    top: 13,
    right: 10,
    bottom: 0,
  },
  buttonStyle: {
    flex: 0,
    marginVertical: 20,
  },
  pdfName: {
    flex: 0.9,
  },
  instruction: {
    color: theme.colors.darkTint3,
    marginBottom: 6,
  },
  selfie: {
    alignSelf: 'center',
    marginVertical: 12,
  },
});
