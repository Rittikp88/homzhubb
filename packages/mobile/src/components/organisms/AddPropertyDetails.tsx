import React, { ReactElement } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { ErrorUtils } from '@homzhub/common/src/utils/ErrorUtils';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { theme } from '@homzhub/common/src/styles/theme';
import { Button, FormTextInput, FurnishingSelection, Text, WithShadowView } from '@homzhub/common/src/components';
import { AssetDescriptionForm } from '@homzhub/mobile/src/components/molecules/AssetDescriptionForm';
import { PropertySpaces } from '@homzhub/mobile/src/components/organisms/PropertySpaces';
import { AssetListingSection } from '@homzhub/mobile/src/components/HOC/AssetListingSection';
import { AssetDescriptionDropdownValues } from '@homzhub/common/src/domain/models/AssetDescriptionForm';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { ISpaceCount, SpaceType } from '@homzhub/common/src/domain/models/AssetGroup';
import { ILastVisitedStep } from '@homzhub/common/src/domain/models/LastVisitedStep';
import { IUpdateAssetParams } from '@homzhub/common/src/domain/repositories/interfaces';

interface IDescriptionForm {
  carpetArea?: number;
  areaUnit?: number;
  buildingGrade?: string;
  facing?: string;
  flooringType?: number;
  yearOfConstruction?: number;
  totalFloors?: number;
  onFloorNumber?: number;
}

interface IFurnishingForm {
  furnishingDetails?: string;
  furnishingType?: string;
}

interface IOwnProps extends WithTranslation {
  assetId: number;
  assetDetails: Asset | null;
  spaceTypes: SpaceType[];
  handleNextStep: () => void;
  lastVisitedStep: ILastVisitedStep;
}

interface IOwnState {
  spacesForm: ISpaceCount[];
  descriptionForm: IDescriptionForm;
  furnishingForm: IFurnishingForm;
  descriptionDropdownValues: AssetDescriptionDropdownValues | null;
}

class AddPropertyDetails extends React.PureComponent<IOwnProps, IOwnState> {
  constructor(props: IOwnProps) {
    super(props);
    const { assetDetails } = this.props;

    this.state = {
      spacesForm: [],
      descriptionForm: {
        carpetArea: (assetDetails && assetDetails.carpetArea) || undefined,
        areaUnit: (assetDetails && assetDetails.carpetAreaUnit && assetDetails.carpetAreaUnit.id) || 1,
        buildingGrade: '',
        facing: (assetDetails && assetDetails.facing) || undefined,
        flooringType: (assetDetails && assetDetails.floorType) || undefined,
        yearOfConstruction: (assetDetails && assetDetails.construction_Year) || undefined,
        totalFloors: (assetDetails && assetDetails.totalFloors) || 0,
        onFloorNumber: (assetDetails && assetDetails.floorNumber) || 0,
      },
      furnishingForm: {
        furnishingDetails: (assetDetails && assetDetails.furnishingDescription) || undefined,
        furnishingType: (assetDetails && assetDetails.furnishing) || undefined,
      },
      descriptionDropdownValues: null,
    };
  }

  public async componentDidMount(): Promise<void> {
    await this.getDescriptionDropdownValues();
  }

  public render(): ReactElement {
    const { spaceTypes, t } = this.props;
    const { descriptionForm, furnishingForm, descriptionDropdownValues } = this.state;

    return (
      <>
        <Formik
          onSubmit={this.onSubmit}
          initialValues={{ ...descriptionForm, ...furnishingForm }}
          validate={FormUtils.validate(this.formSchema)}
        >
          {(formProps: FormikProps<FormikValues>): React.ReactNode => {
            return (
              <>
                <View style={styles.containerStyle}>
                  <>
                    <Text style={styles.headingStyle} type="small">
                      {t('property:spacesText')}
                    </Text>
                    <PropertySpaces onChange={this.handleSpaceFormChange} spacesTypes={spaceTypes} />
                  </>

                  {descriptionDropdownValues && (
                    <AssetDescriptionForm dropDownOptions={descriptionDropdownValues} formProps={formProps} />
                  )}

                  {descriptionDropdownValues && this.renderFurnishingFields(formProps)}
                </View>

                <WithShadowView>
                  <Button
                    type="primary"
                    title={t('common:continue')}
                    containerStyle={styles.buttonStyle}
                    onPress={(): Promise<void> => this.onSubmit(formProps.values)}
                  />
                </WithShadowView>
              </>
            );
          }}
        </Formik>
      </>
    );
  }

  private renderFurnishingFields = (formProps: FormikProps<FormikValues>): ReactElement => {
    const { t } = this.props;
    const onFurnishingChange = (value: string): void => this.setFurnishingStatus(formProps, value);

    return (
      <AssetListingSection title={t('property:furnishing')} containerStyles={styles.furnishingStyle}>
        <>
          <FurnishingSelection
            titleHidden
            value={formProps.values.furnishingType}
            onFurnishingChange={onFurnishingChange}
          />
          <FormTextInput
            style={styles.furnishingFieldStyle}
            name="furnishingDetails"
            label={t('property:furnishingDetails')}
            maxLength={400}
            inputType="default"
            formProps={formProps}
          />
        </>
      </AssetListingSection>
    );
  };

  private onSubmit = async (values: FormikValues): Promise<void> => {
    const {
      areaUnit,
      carpetArea,
      facing,
      flooringType,
      furnishingDetails,
      onFloorNumber,
      totalFloors,
      yearOfConstruction,
      furnishingType,
    } = values;
    const { spacesForm } = this.state;
    const { handleNextStep, assetId, lastVisitedStep } = this.props;

    const sanitizedSpaces = spacesForm
      .filter((item) => item && item.description !== '')
      .map((space) => {
        return {
          space_type: space.space_type,
          count: space.count,
          ...(space.description && { description: space.description }),
        };
      });

    const payload: IUpdateAssetParams = {
      carpet_area: carpetArea,
      carpet_area_unit: areaUnit,
      facing,
      floor_type: flooringType,
      furnishing_description: furnishingDetails,
      floor_number: onFloorNumber,
      total_floors: totalFloors,
      construction_year: yearOfConstruction,
      furnishing: furnishingType,
      spaces: sanitizedSpaces,
      last_visited_step: {
        ...lastVisitedStep,
        asset_creation: {
          ...lastVisitedStep.asset_creation,
          is_details_done: true,
          total_step: 4,
        },
      },
    };

    try {
      await AssetRepository.updateAsset(assetId, payload);
      handleNextStep();
    } catch (e) {
      const error = ErrorUtils.getErrorMessage(e.details);
      AlertHelper.error({ message: error });
    }
  };

  private setFurnishingStatus = (formProps: FormikProps<FormikValues>, furnishingType: string): void => {
    formProps.setFieldValue('furnishingType', furnishingType);
  };

  private handleSpaceFormChange = (id: number, count: number, description?: string): void => {
    const { spacesForm } = this.state;

    spacesForm[id] = { space_type: id, count, description };
  };

  private formSchema = (): yup.ObjectSchema<IDescriptionForm & IFurnishingForm> => {
    return yup.object().shape({
      carpetArea: yup.number().optional(),
      areaUnit: yup.number().optional(),
      buildingGrade: yup.string().optional(),
      facing: yup.string().optional(),
      flooringType: yup.number().optional(),
      yearOfConstruction: yup.number().optional(),
      totalFloors: yup.number().optional(),
      onFloorNumber: yup.number().optional(),
      furnishingType: yup.string().optional(),
      furnishingDetails: yup.string().optional(),
    });
  };

  private getDescriptionDropdownValues = async (): Promise<void> => {
    try {
      const response: AssetDescriptionDropdownValues = await AssetRepository.getAssetDescriptionDropdownValues();

      this.setState({
        descriptionDropdownValues: response,
        furnishingForm: {
          furnishingType: response.furnishingStatus[0].name,
        },
      });
    } catch (e) {
      AlertHelper.error({ message: e.message });
    }
  };
}

const addPropertyDetails = withTranslation()(AddPropertyDetails);
export { addPropertyDetails as AddPropertyDetails };

const styles = StyleSheet.create({
  containerStyle: {
    paddingHorizontal: theme.layout.screenPadding,
  },
  buttonStyle: {
    flex: 0,
    margin: 16,
  },
  furnishingStyle: {
    marginTop: 16,
  },
  headingStyle: {
    marginTop: 16,
    paddingVertical: 16,
    paddingLeft: 16,
    backgroundColor: theme.colors.moreSeparator,
  },
  furnishingFieldStyle: {
    height: 85,
    textAlignVertical: 'top',
    paddingTop: 10,
  },
});
