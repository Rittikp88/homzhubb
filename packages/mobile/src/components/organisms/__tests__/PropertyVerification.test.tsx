import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyVerification } from '@homzhub/mobile/src/components/organisms/PropertyVerification';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { PropertyVerificationTypes } from '@homzhub/common/src/mocks/PropertyVerification';

const mock = jest.fn();

describe('Property Verification', () => {
  let component: ShallowWrapper;
  let props: any;
  let instance: any;

  beforeEach(() => {
    props = {
      navigateToPropertyHelper: mock,
      typeOfFlow: 'RENT',
      updateStep: mock,
      propertyId: 63,
    };
    component = shallow(<PropertyVerification {...props} t={(key: string): string => key} />);
    instance = component.instance();
  });

  it('should match snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should fetch the verification document list', async () => {
    jest
      .spyOn(AssetRepository, 'getVerificationDocumentTypes')
      .mockImplementation(() => Promise.resolve(PropertyVerificationTypes));
    const existingDocuments = await AssetRepository.getVerificationDocumentTypes();
    instance.setState({ verificationTypes: existingDocuments });
  });
});
