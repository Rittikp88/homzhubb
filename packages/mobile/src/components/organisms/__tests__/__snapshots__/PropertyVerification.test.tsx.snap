// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyVerification should match snapshot 1`] = `
<Fragment>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "marginTop": 4,
        "paddingHorizontal": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        key="0"
        style={
          Object {
            "marginBottom": 10,
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4C5F72",
            }
          }
          textType="semiBold"
          type="small"
        >
          ID Proof
        </Text>
        <UploadBox
          containerStyle={
            Object {
              "marginTop": 20,
            }
          }
          header="Upload ID proof"
          icon="id-proof"
          onPress={[Function]}
        />
      </View>
      <View
        key="1"
        style={
          Object {
            "marginBottom": 10,
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4C5F72",
            }
          }
          textType="semiBold"
          type="small"
        >
          Click a selfie with the ID proof
        </Text>
        <[object Object]
          style={
            Object {
              "alignSelf": "center",
              "marginVertical": 12,
            }
          }
        />
        <Label
          key="0"
          style={
            Object {
              "color": "#33495E",
              "marginBottom": 6,
            }
          }
          textType="regular"
          type="regular"
        >
          1. Make sure your face is clearly visible
        </Label>
        <Label
          key="1"
          style={
            Object {
              "color": "#33495E",
              "marginBottom": 6,
            }
          }
          textType="regular"
          type="regular"
        >
          2. Please ensure that the ID is clearly visible in your photo
        </Label>
        <UploadBox
          containerStyle={
            Object {
              "marginTop": 20,
            }
          }
          header="Capture Your Selfie"
          icon="selfie"
          onPress={[Function]}
        />
      </View>
      <View
        key="2"
        style={
          Object {
            "marginBottom": 10,
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4C5F72",
            }
          }
          textType="semiBold"
          type="small"
        >
          Upload Ownership Verification Document
        </Text>
        <Label
          style={
            Object {
              "color": "#33495E",
              "marginVertical": 10,
            }
          }
          textType="regular"
          type="regular"
        >
          You can upload Electricity bill or Property tax reciepts which mentions your name and the property address
        </Label>
        <UploadBox
          containerStyle={
            Object {
              "marginTop": 20,
            }
          }
          header="Upload Document"
          icon="document"
          onPress={[Function]}
        />
      </View>
      <View
        key="3"
        style={
          Object {
            "marginBottom": 10,
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4C5F72",
            }
          }
          textType="semiBold"
          type="small"
        >
          Upload Occupancy Document
        </Text>
        <Label
          style={
            Object {
              "color": "#33495E",
              "marginVertical": 10,
            }
          }
          textType="regular"
          type="regular"
        >
          You can upload Electricity bill or Property tax reciepts which mentions your name and the property address
        </Label>
        <UploadBox
          containerStyle={
            Object {
              "marginTop": 20,
            }
          }
          header="Upload Document"
          icon="document"
          onPress={[Function]}
        />
      </View>
      <View
        key="4"
        style={
          Object {
            "marginBottom": 10,
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4C5F72",
            }
          }
          textType="semiBold"
          type="small"
        >
          Upload Property Tax
        </Text>
        <Label
          style={
            Object {
              "color": "#33495E",
              "marginVertical": 10,
            }
          }
          textType="regular"
          type="regular"
        >
          You can upload Electricity bill or Property tax reciepts which mentions your name and the property address
        </Label>
        <UploadBox
          containerStyle={
            Object {
              "marginTop": 20,
            }
          }
          header="Upload Document"
          icon="document"
          onPress={[Function]}
        />
      </View>
    </View>
  </View>
  <Button
    containerStyle={
      Object {
        "flex": 0,
        "marginTop": 20,
      }
    }
    disabled={true}
    onPress={[Function]}
    title="common:continue"
    type="primary"
  />
</Fragment>
`;
